### PR TITLE
Develop deprecate broadcasting

### PR DESCRIPTION
**Pull request by vincentpierre**
[Associated Design Document](https://docs.google.com/document/d/1q_eJ5KHNcVXoEBkGxTm-3Kt4zl4_aAoeuccrdIoEfkI/edit#)

Builds on top of #2659 

In this PR : 
 - Removal of the `Control` checkbox in the BroadcastHub of the Academy
 - By default, all learning brains will be trained by Python
 - The BroadcastHub is now a list of **LearningBrains** and not **Brains**
 - In Editor : 
    - If no brains are in the Hub OR if there is no Python process to communicate with Inference
    - If there is at least one Brain in the Hub and UnityEnvironment was created without an executable : Training
 - In Executable : 
    - If launched with a UnityEnvironment : Training
    - If opened manually : Inference

TODO : 
 - Documentation
 - Review code comments

### CODE REVIEW

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 1]
```diff
@@ -284,6 +277,7 @@ private void InitializeEnvironment()
             catch
             {
                 communicator = null;
+#if UNITY_EDITOR
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Can you update the comments above to include the change in logic?

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 2]
```diff
@@ -323,9 +316,20 @@ private void InitializeEnvironment()
                         key, resetParameters[key]
                     );
                 }
-
-                var pythonParameters = m_BrainBatcher.SendAcademyParameters(academyParameters);
-                Random.InitState(pythonParameters.Seed);
+                try
+                {
+                    var pythonParameters = m_BrainBatcher.SendAcademyParameters(academyParameters);
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: What causes this to raise an exception? Please explain the logic around here more.

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 3]
```diff
@@ -323,9 +316,20 @@ private void InitializeEnvironment()
...
+                    Random.InitState(pythonParameters.Seed);
+                }
+                catch
+                {
+                    communicator = null;
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: The batcher still has a reference to the (now defunct) communicator, right?

Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Yes, you are right. I overwrite the batcher now.

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 2 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: I added a comment

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 1 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Updated comment

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 3 above)
Comment (chriselion) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Sorry, missed that below.

File: UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs [hunk 4]
```diff
@@ -11,50 +11,27 @@ namespace MLAgents
     [System.Serializable]
```
Comment (surfnerd) [inline UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs]: is there any need for this class anymore?

Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs]: Yes. Still needs more work to kill it completely. You must put your Brains into the Hub so the Academy knows what brains are training.

File: UnitySDK/Assets/ML-Agents/Examples/3DBall/Scenes/3DBall.unity [hunk 5]
```diff
@@ -851,11 +851,10 @@ MonoBehaviour:
   m_Script: {fileID: 11500000, guid: eb15e3c3d55e54abaafb74c635b6a458, type: 3}
   m_Name: 
   m_EditorClassIdentifier: 
+  trainInEditor: 0
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Examples/3DBall/Scenes/3DBall.unity]: Is this used anywhere? Don't see it referenced in the code.

File: UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs [hunk 6]
```diff
@@ -11,50 +11,27 @@ namespace MLAgents
...
     public class BroadcastHub
     {
-        [SerializeField]
-        public List<Brain> broadcastingBrains = new List<Brain>();
         [FormerlySerializedAs("_brainsToControl")]
         [SerializeField]
-        private List<Brain> m_BrainsToControl = new List<Brain>();
+        public List<LearningBrain> brainsToControl = new List<LearningBrain>();
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs]: Should we make this `[FormerlySerializedAs("broadcastingBrains"]`? or is that going to break loading old assets? What's the upgrade path for users?

File: UnitySDK/Assets/ML-Agents/Examples/3DBall/Scenes/3DBall.unity (hunk 5 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Examples/3DBall/Scenes/3DBall.unity]: Nope, not used

File: UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs (hunk 6 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs]: I updated all the scenes.

Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/BroadcastHub.cs]: Maybe I can just remove the FormerlySerializedAs. It was added automatically by my IDE

File: UnitySDK/Assets/ML-Agents/Examples/3DBall/Scenes/3DBall.unity (hunk 5 above)
Comment (chriselion) [reply UnitySDK/Assets/ML-Agents/Examples/3DBall/Scenes/3DBall.unity]: It's not just here, it's in (almost?) all the other modified assets too.

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 7]
```diff
@@ -94,12 +72,11 @@ private void Shutdown()
         /// </summary>
         private void BrainDecideAction()
         {
-            m_BrainBatcher?.SendBrainInfo(name, m_AgentInfos);
             DecideAction();
         }
 
         /// <summary>
-        /// Is called only once at the begening of the training or inference session.
+        /// Is called only once at the begining of the training or inference session.
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: nit: "beginning" :P

**Pull request merged**